[irods/irods#8257] Add migration guide for legacy auth plugins (main) (#344)
irods/irods_docs: docs/plugins/pluggable_authentication.md

Over time, the legacy authentication plugins fell out of use in the Consortium and larger iRODS community. In 4.3, the only supported legacy authentication plugins were iRODS native authentication and PAM.

In iRODS 5, the legacy authentication plugins were declared deprecated. The legacy authentication plugins will be removed in iRODS 6.

### How to know if a client is using legacy authentication plugins

!!! Note
    iCommands with version 4.3.0 or later do NOT use the deprecated legacy authentication plugins.

In iRODS 5, the server-side legacy authentication plugin operations for the only remaining supported legacy authentication plugins return a message on the `rError` stack as well as a `DEPRECATED_AUTHENTICATION_PLUGIN` error code (-3002000). The error code is not indicative of an error occurring with the authentication, but is used to signal to the caller that the deprecated legacy authentication plugins are being used. If your client supports the `rError` stack, you should see a message like this:

> The legacy authentication plugin interface has been deprecated. Please update client to not use legacy authentication plugins.

> If your client does not support `rError` stack, or you are not sure how to see its contents, we will need to examine the source code. The legacy authentication API endpoints have been deprecated as well. If any of the following API numbers are used by your client library for authentication, it is using legacy authentication plugins:

Reviewer: ```suggestion
If your client does not support the `rError` stack, or you are not sure how to see its contents, we will need to examine the source code. The legacy authentication API endpoints have been deprecated as well. If any of the following API numbers are used by your client library for authentication, it is using legacy authentication plugins:
```

Author: Done

- `AUTH_REQUEST_AN` (703)
- `AUTH_RESPONSE_AN` (704)
- `PAM_AUTH_REQUEST_AN` (725)
- `AUTH_PLUG_REQ_AN` (1201)
- `AUTH_PLUG_RESP_AN` (1202)

### Implementing the authentication plugin framework

!!! Note
    This section will likely only be of interest to client library developers.

The authentication plugin framework is client-driven, so each client library must maintain an implementation of the client-side plugin operations for each authentication scheme. The legacy authentication plugins also required each client library to implement all client-side operations for each authentication scheme, so this is no different in principle.

> Client libraries are free to implement client-side authentication operations in whatever way they please. However, in order to use the authentication schemes supported by a given server, the server-side authentication operations must be called with the expected data payload at the expected times. Down below, we will explain how the supported authentication schemes operate in turn.

Reviewer: ```suggestion
Client libraries are free to implement client-side authentication operations in whatever way they please. However, in order to use the authentication schemes supported by a given server, the server-side authentication operations must be called with the expected data payload at the expected times. Below, we will explain how the supported authentication schemes operate in turn.
```

Author: Done

#### Client-driven authentication flow

The C++ implementation for the authentication plugin framework follows these steps to authenticate a client:

1. Load the shared object for the appropriate plugin based on the authentication scheme in the client environment.

> 2. Call "start" client operation for that plugin.

Reviewer: ```suggestion
2. Call the "start" client operation for that plugin.
```

Author: Done

### Implementing authentication plugins

!!! Note
    This section will likely only be of interest to client library developers.

#### Native authentication

The native authentication plugin has two server-side operations which must be called in order. The inputs and outputs from each are described below.

**Step 1: `auth_agent_auth_request`**

This operation - like all server-side plugin operations - requires a "scheme" in the request payload so that the server knows which authentication plugin to load. Otherwise, this operation has no requirements in terms of keys to include in the request payload. The request payload should minimally contain the following:
```json
{
    "scheme": "native"
}
```

The response payload from the server will minimally contain the following:
```json
{
    "scheme": "native",
    "request_result": "64_random_bytes_encoded_as_a_string____________________________"
}
```
The `request_result` is a randomly-generated byte-string that the server also preserves and uses to verify the client with which it is communicating.

**Step 2: `auth_agent_auth_response`**

This operation requires 3 keys: `digest`, `user_name`, and `zone_name`.

> `user_name` and `zone_name` are the username and zone name for the user being authenticated. A user of the specified name belonging the specified zone must exist in the local zone in order for authentication to succeed.

Reviewer: ```suggestion
`user_name` and `zone_name` are the username and zone name for the user being authenticated. A user of the specified name belonging to the specified zone must exist in the local zone's catalog in order for authentication to succeed.
```

`digest` is a base64-encoded MD5 hash of the user's password. The client should construct the MD5 hash by concatenating two strings:

1. The randomly-generated string from the server stored in `request_result`
2. The user's plaintext password

How the plugin gets the user's password is entirely up to the library developer. For reference, the C++ client-side authentication plugin has a few ways of obtaining the password:

- The `a_pw` key in the request payload

> - Prompting the user and receiving the passwords from `stdin`

Reviewer: ```suggestion
- Prompting the user and receiving a password from `stdin`
```

Author: Done

> - Reading from an irodsA file

Reviewer: ```suggestion
- Reading from a `.irodsA` file
```

Author: Done

For example, if the user's password is "apass", the string to hash looks like this:
```
64_random_bytes_encoded_as_a_string____________________________apass
```
The concatenated string should be used as input to a MD5 hashing function, and the resulting byte-string should be base64-encoded. The resulting string should be stored in the `digest` key.

The request payload should minimally contain the following:
```json
{
    "scheme": "native",
    "user_name": "<user_name>",
    "zone_name": "<zone_name>",

>     "digest": "<base64-encoded string of MD5 hash>"

Reviewer 2: Consider replacing the hyphens and spaces with underscores in the value for `digest`. Or remove the underscores from the user name and zone name values.

Author: Removed underscores

}
```

The response payload from the server will be identical to the request as the server-side operation does not add or remove anything.

After `auth_agent_auth_response` completes, the `RsComm` in the connected agent has all of the appropriate values set and should be ready to perform regular requests.

#### PAM (`pam_password`) authentication

The `pam_password` authentication plugin has one server-side operation which must be called. The inputs and outputs are described below.

After this server-side operation returns, the client is expected to authenticate again using a randomly-generated iRODS native authentication password.

**Step 1: `auth_agent_auth_request`**

This operation requires 3 keys: `a_pw`, `user_name`, and `zone_name`.

> `user_name` and `zone_name` are the username and zone name for the user being authenticated. A user of the specified name belonging the specified zone must exist in the local zone in order for authentication to succeed.

Reviewer: ```suggestion
`user_name` and `zone_name` are the username and zone name for the user being authenticated. A user of the specified name belonging to the specified zone must exist in the local zone's catalog in order for authentication to succeed.
```

Author: Done

> `a_pw` is the user's **plaintext PAM password** which is going to be checked by the configured PAM service. This is why communication between client and server requires TLS to be enabled.

Reviewer: ```suggestion
`a_pw` is the user's **plaintext PAM password** which is going to be checked by the configured server-side PAM service. This is why communication between client and server requires TLS to be enabled.
```

Author: Done

The request payload should minimally contain the following:
```json
{
    "scheme": "pam_password",
    "user_name": "<user_name>",
    "zone_name": "<zone_name>",

>     "a_pw": "<plaintext PAM password>"

Reviewer 2: Consider replacing the hyphens and spaces with underscores in the value for `a_pw`. Or remove the underscores from the user name and zone name values.

Author: Removed underscores

}
```

The response payload from the server will minimally contain the following:
```json
{
    "scheme": "pam_password",
    "user_name": "<user_name>",
    "zone_name": "<zone_name>",

>     "request_result": "<randomly-generated iRODS password>"

Reviewer 2: Consider replacing the hyphens and spaces with underscores in the value for `request_result`. Or remove the underscores from the user name and zone name values.

}
```

> The `request_result` key contains a time-limited, randomly-generated iRODS password. This password should be used to authenticate using the native authentication scheme. How the client implements this authentication is entirely up to the developer. For reference, the C++ `pam_password` client-side authentication plugin saves the randomly-generated password to an irodsA file and then starts the authentication process from the "start" operation using the native authentication scheme and the password in that irodsA file.

Reviewer: ```suggestion
The `request_result` key contains a time-limited, randomly-generated iRODS password. This password should be used to authenticate using the native authentication scheme. How the client implements this authentication is entirely up to the developer. For reference, the C++ `pam_password` client-side authentication plugin saves the randomly-generated password to a `.irodsA` file and then starts the authentication process from the "start" operation using the native authentication scheme and the password in that `.irodsA` file.
```

Author: Done
